File: src/lib/translation.ts
const rawBase = import.meta.env.VITE_API_URL as string | undefined;
const API_BASE_URL = (rawBase || 'http://localhost:8000').replace(/^\"|\"$/g, '').trim();

const FALLBACK_LANGUAGES: Record<string, string> = {
  en: 'English',
  hi: 'Hindi',
  mr: 'Marathi',
  ta: 'Tamil',
  te: 'Telugu',
  bn: 'Bengali',
  gu: 'Gujarati',
  kn: 'Kannada',
  ml: 'Malayalam',
  pa: 'Punjabi',
};

// Helper function to handle API requests
async function apiRequest<T>(url: string, options?: RequestInit): Promise<T> {
  try {
    const controller = new AbortController();
    const id = setTimeout(() => controller.abort(), 2000);
    const response = await fetch(url, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...(options?.headers || {})
      },
      signal: controller.signal,
    });
    clearTimeout(id);

    if (!response.ok) {
      throw new Error(`API request failed with status ${response.status}`);
    }

    return await response.json();
  } catch (error) {
    // Downgrade to warn to avoid noisy console when backend is not running
    console.warn('[translation] API request error or timeout:', error);
    throw error;
  }
}

interface TranslationResult {
  original_text: string;
  transliterated_text: string;
}

interface TranslationResponse {
  status: string;
  data: TranslationResult;
}

interface LanguageListResponse {
  status: string;
  languages: Record<string, string>;
}

export const translationService = {
  /**
   * Get list of supported languages
   */
  async getSupportedLanguages(): Promise<Record<string, string>> {
    try {
      const data = await apiRequest<LanguageListResponse>(
        `${API_BASE_URL}/api/translation/languages`
      );
      return data.languages || {};
    } catch (error) {
      console.warn('[translation] Backend unavailable, using fallback languages');
      return FALLBACK_LANGUAGES;
    }
  },

  /**
   * Transliterate a single word
   */
  async transliterateWord(
    text: string,
    targetLang: string = 'hi'
  ): Promise<TranslationResult | null> {
    if (!text.trim()) return null;

    try {
      const params = new URLSearchParams({
        text,
        lang: targetLang,
      });
      
      const url = `${API_BASE_URL}/api/translation/transliterate/word?${params}`;
      const data = await apiRequest<TranslationResponse>(url);
      return data.data;
    } catch (error) {
      console.warn('[translation] Fallback transliteration (echo)');
      return { original_text: text, transliterated_text: text };
    }
  },

  /**
   * Transliterate a sentence
   */
  async transliterateSentence(
    text: string,
    targetLang: string = 'hi'
  ): Promise<TranslationResult | null> {
    if (!text.trim()) return null;

    try {
      const params = new URLSearchParams({
        text,
        lang: targetLang,
      });
      
      const url = `${API_BASE_URL}/api/translation/transliterate/sentence?${params}`;
      const data = await apiRequest<TranslationResponse>(url);
      return data.data;
    } catch (error) {
      console.warn('[translation] Fallback transliteration (echo)');
      return { original_text: text, transliterated_text: text };
    }
  },

  /**
   * Get language name from code
   */
  getLanguageName(
    code: string,
    languages: Record<string, string>
  ): string {
    return languages[code] || code;
  },
};

File: backend/app/api/endpoints/translation.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from typing import Optional
import logging

from app.services.translation_service import translation_service

router = APIRouter()
logger = logging.getLogger(__name__)

@router.get("/languages")
async def get_supported_languages():
    """Get list of supported languages for transliteration"""
    try:
        languages = translation_service.get_supported_languages()
        return {
            "status": "success",
            "languages": languages
        }
    except Exception as e:
        logger.error(f"Error getting supported languages: {str(e)}")
        raise HTTPException(status_code=500, detail="Failed to retrieve supported languages")

@router.get("/transliterate/word")
async def transliterate_word(
    text: str,
    target_lang: str = 'hi',
):
    """
    Transliterate a single word from English to the target language script
    
    Args:
        text: The English text to transliterate
        target_lang: Target language code (default: 'hi' for Hindi)
    """
    try:
        if not text.strip():
            raise HTTPException(status_code=400, detail="Text cannot be empty")
            
        result = await translation_service.transliterate_word(text, target_lang)
        return {
            "status": "success",
            "data": result
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in transliteration: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Translation failed: {str(e)}")

@router.get("/transliterate/sentence")
async def transliterate_sentence(
    text: str,
    target_lang: str = 'hi',
):
    """
    Transliterate a sentence from English to the target language script
    
    Args:
        text: The English text to transliterate
        target_lang: Target language code (default: 'hi' for Hindi)
    """
    try:
        if not text.strip():
            raise HTTPException(status_code=400, detail="Text cannot be empty")
            
        result = await translation_service.transliterate_sentence(text, target_lang)
        return {
            "status": "success",
            "data": result
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Error in sentence transliteration: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Translation failed: {str(e)}")
